### PR TITLE
[Feat](Nereids) join hint support stage one

### PR DESCRIPTION
**Pull request by LiBinfeng-01**
## Proposed changes

- support view as a independent unit of leading hint
- add random test check of leading hint query
- add more test with data of leading hint query
- add random test check of distribute hint

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Review by englefly — DISMISSED** [review]

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateLogicalSelectHint.java [hunk 1]
```diff
@@ -15,61 +15,65 @@
...
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
+import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Optional;
 
 /**
- * eliminate set var hint, and set var to session variables.
+ * eliminate logical select hint and set then to cascade context
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateLogicalSelectHint.java:46]: "then"->"them"

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateLogicalSelectHint.java [hunk 2]
```diff
@@ -15,61 +15,65 @@
...
+                    }
+                    OrderedHint ordered = new OrderedHint("Ordered");
+                    ordered.setStatus(Hint.HintStatus.SUCCESS);
+                    ctx.cascadesContext.getHintMap().put("Ordered", ordered);
+                    ctx.statementContext.addHint(ordered);
+                } else if (hintName.equalsIgnoreCase("LEADING")) {
+                    extractLeading((SelectHintLeading) hint.getValue(), ctx.cascadesContext,
+                            ctx.statementContext, selectHintPlan.getHints());
+                } else {
+                    // logger.warn("Can not process select hint '{}' and skip it", hint.getKey());
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateLogicalSelectHint.java:72]: it is better to  open this comments

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateLogicalSelectHint.java (hunk 1 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateLogicalSelectHint.java:46]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateLogicalSelectHint.java (hunk 2 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateLogicalSelectHint.java:72]: done

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java [hunk 3]
```diff
@@ -367,6 +435,13 @@ public Plan generateLeadingJoinPlan() {
                             newStackTop.second.getOutput(), logicalPlan.getOutput(), conditions);
                     JoinType joinType = computeJoinType(getBitmap(newStackTop.second),
                             getBitmap(logicalPlan), conditions);
+                    if (joinType == null) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:438]: better to refine this part of return type handing.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java [hunk 4]
```diff
@@ -191,6 +196,29 @@ public List<Pair<Long, Expression>> getFilters() {
...
+    }
+
+    /**
+     * find out whether filters can match original joinType
+     * @param filters filters needs to put on this join
+     * @param joinType join type computed by join constraint
+     * @return can filters matched
+     */
+    public boolean isFilterJoinTypeMatched(List<Expression> filters, JoinType joinType) {
+        for (Expression filter : filters) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:210]: Pls confirm the definition of filter and join condition.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java [hunk 5]
```diff
@@ -191,6 +196,29 @@ public List<Pair<Long, Expression>> getFilters() {
...
+            JoinType originalJoinType = filterJoinType.get(filter);
+            if (originalJoinType.equals(joinType)
+                    || originalJoinType.isOuterJoin() && joinType.isOuterJoin()
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:213]: Pls confirm the isOuterJoin scope is too big. Consider the original type is full outer and leading is left join.

Comment (xzj7019) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:213]: Pls also check if it is covered by random test and current ut.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java [hunk 6]
```diff
@@ -401,17 +482,6 @@ public Plan generateLeadingJoinPlan() {
 
         LogicalJoin finalJoin = (LogicalJoin) stack.pop().second;
         // we want all filters been remove
-        if (!getFilters().isEmpty()) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:404]: Pls add a checking for getFilters is not empty.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java [hunk 7]
```diff
@@ -121,6 +124,12 @@ private static List<RewriteJob> buildAnalyzeJobs(Optional<CustomTableResolver> c
             bottomUp(new CheckAnalysis()),
             topDown(new EliminateGroupByConstant()),
             topDown(new NormalizeAggregate()),
+            bottomUp(new SemiJoinCommute()),
+            bottomUp(
+                    new CollectSubQueryAlias(),
+                    new CollectJoinConstraint()
+            ),
+            topDown(new LeadingJoin()),
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java:132]: Pls move this rule from rewrite stage to analyze stage.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java [hunk 8]
```diff
@@ -121,6 +124,12 @@ private static List<RewriteJob> buildAnalyzeJobs(Optional<CustomTableResolver> c
             bottomUp(new CheckAnalysis()),
             topDown(new EliminateGroupByConstant()),
             topDown(new NormalizeAggregate()),
+            bottomUp(new SemiJoinCommute()),
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java:127]: Pls add a TODO here to rename SemiJoinCommute() to JoinCommute()

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CollectJoinConstraint.java [hunk 9]
```diff
@@ -47,24 +50,17 @@ public class CollectJoinConstraint implements RewriteRuleFactory {
...
-                LeadingHint leading = (LeadingHint) ctx.cascadesContext
-                        .getStatementContext().getHintMap().get("Leading");
-                if (leading == null) {
+                if (!ctx.cascadesContext.isLeadingJoin()) {
                     return ctx.root;
                 }
+                LeadingHint leading = (LeadingHint) ctx.cascadesContext
+                            .getHintMap().get("Leading");
                 LogicalJoin join = ctx.root;
+                if (join.getJoinType().isNullAwareLeftAntiJoin()) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CollectJoinConstraint.java:60]: Pls add this restriction into our document tracking.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CollectJoinConstraint.java [hunk 10]
```diff
@@ -92,27 +90,37 @@ public List<Rule> buildRules() {
...
+                if (!ctx.cascadesContext.isLeadingJoin()) {
                     return ctx.root;
                 }
+                LeadingHint leading = (LeadingHint) ctx.cascadesContext
+                        .getHintMap().get("Leading");
                 LogicalFilter filter = ctx.root;
                 Set<Expression> expressions = filter.getConjuncts();
                 for (Expression expression : expressions) {
-                    Long filterBitMap = calSlotsTableBitMap(leading, expression.getInputSlots(), false);
+                    Long filterBitMap = 0L;
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CollectJoinConstraint.java:101]: Pls make sure this part of logic correctness and completeness.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java (hunk 5 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:213]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java (hunk 4 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:210]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java (hunk 6 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:404]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java (hunk 7 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java:132]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java (hunk 8 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java:127]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CollectJoinConstraint.java (hunk 9 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CollectJoinConstraint.java:60]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java (hunk 3 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/hint/LeadingHint.java:438]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CollectJoinConstraint.java (hunk 10 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CollectJoinConstraint.java:101]: done

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Review by englefly — APPROVED** [review]

**Review by xzj7019 — APPROVED** [review]

**Pull request merged**